libraries/AP_Math/tests/test_math.cpp

```diff
@@ -0,0 +1,309 @@
+{  
+  for(int i = 0; i < 1000; i++) {
+      if(i < 250) {
+          EXPECT_EQ(constrain_float(i, 250, 500), 250);
+          EXPECT_EQ(constrain_int16(i, 250, 500), 250);
+          EXPECT_EQ(constrain_int32(i, 250, 500), 250);
+      }
+      else if(i > 500) {
+          EXPECT_EQ(constrain_float(i, 250, 500), 500);
+          EXPECT_EQ(constrain_int16(i, 250, 500), 500);
+          EXPECT_EQ(constrain_int32(i, 250, 500), 500);
+      }
+      else {
+          EXPECT_EQ(constrain_float(i, 250, 500), i);        
+          EXPECT_EQ(constrain_int16(i, 250, 500), i);
+          EXPECT_EQ(constrain_int32(i, 250, 500), i);
+      }
+  }
+  
+  for(int i = 0; i <= 1000; i++) {
+      int c = i - 1000;
+      if(c < -250) {
+          EXPECT_EQ(constrain_float(c, -250, -50), -250);
+          EXPECT_EQ(constrain_int16(c, -250, -50), -250);
+          EXPECT_EQ(constrain_int32(c, -250, -50), -250);
+      }
+      else if(c > -50) {
+          EXPECT_EQ(constrain_float(c, -250, -50), -50);
+          EXPECT_EQ(constrain_int16(c, -250, -50), -50);
+          EXPECT_EQ(constrain_int32(c, -250, -50), -50);
+      }
+      else {
+          EXPECT_EQ(constrain_float(c, -250, -50), c);        
+          EXPECT_EQ(constrain_int16(c, -250, -50), c);
+          EXPECT_EQ(constrain_int32(c, -250, -50), c);
+      }
+  }
+  
+  for(int i = 0; i <= 2000; i++) {
+      int c = i - 1000;
+      if(c < -250) {
+          EXPECT_EQ(constrain_float(c, -250, 50), -250);
+          EXPECT_EQ(constrain_int16(c, -250, 50), -250);
+          EXPECT_EQ(constrain_int32(c, -250, 50), -250);
+      }
+      else if(c > 50) {
+          EXPECT_EQ(constrain_float(c, -250, 50), 50);
+          EXPECT_EQ(constrain_int16(c, -250, 50), 50);
+          EXPECT_EQ(constrain_int32(c, -250, 50), 50);
+      }
+      else {
+          EXPECT_EQ(constrain_float(c, -250, 50), c);        
+          EXPECT_EQ(constrain_int16(c, -250, 50), c);
+          EXPECT_EQ(constrain_int32(c, -250, 50), c);
+      }
+  }
+}
+
+TEST(Wrap180Test, angle180)
+{
+  // Full circle test
+  for(int i = 0; i < 360; i++) {
+      // smaller pole position
+      if(i < 180) {
+          EXPECT_EQ(wrap_180(i), i);
+          EXPECT_EQ(wrap_180(-i), -i);
+      }
+      // hit pole position -180/+180 degree
+      else if(i == 180) {
+          EXPECT_EQ(wrap_180(i), i);
+          EXPECT_EQ(wrap_180(-i), i);
+      }
+      // bigger pole position
+      else {
+          EXPECT_EQ(wrap_180(i), -(360-i));
+          EXPECT_EQ(wrap_180(-i), (360-i));
+      }
+  }
+  
+  EXPECT_EQ(wrap_180(45.f),        45.f);
+  EXPECT_EQ(wrap_180(90.f),        90.f);
+  EXPECT_EQ(wrap_180(180.f),       180.f);
+  EXPECT_EQ(wrap_180(180.1f),     -179.9f);
+  EXPECT_EQ(wrap_180(270.f),      -90.f);
+  EXPECT_EQ(wrap_180(360.f),       0.f);
+  EXPECT_EQ(wrap_180(720.f),       0.f);
+  EXPECT_EQ(wrap_180(3600.f),      0.f);
+  EXPECT_EQ(wrap_180(7200.f),      0.f);
+  EXPECT_EQ(wrap_180(-36000000.f), 0.f);
```

Comment:
A good catch. A temp is created which adds 18000 to the input angle and leads I guess to loss of floating point precision
Suggest to implement wrap180 in terms of wrap360
float wrap_180(v ,scale)
{
   auto res = wrap_360(v, scale);
  if ( res > 180 * scale)...  res -= 180 * scale
}
HTH. Sorry dont have gtest framework installed to test properly. Will try to get around to it